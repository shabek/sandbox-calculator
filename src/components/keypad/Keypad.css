.keypad {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  gap: 8px;
  width: 100%;
}

.keypad-btn {
  align-items: center;
  border: 1px solid #f3f3f3;
  border-radius: 0.5rem;
  background-color: #f3f3f3;
  color: #000;
  cursor: pointer;
  display: flex;
  font-family: inherit;
  font-size: 1rem;
  justify-content: center;
  padding: 0.5rem;
  width: 100%;
}

.keypad-btn.primary {
  border-color: #4285f4;
  background-color: #4285f4;
  color: #fff;
  font-weight: 700;
}

.keypad-btn.dark {
  border-color: #d3d3d3;
  background-color: #d3d3d3;
}

.keypad-btn.wide {
  grid-column: span 2;
}

.keypad-btn:focus,
.keypad-btn.dark:focus,
.keypad-btn.primary:focus {
  border-color: #000;
  outline: none;
}
