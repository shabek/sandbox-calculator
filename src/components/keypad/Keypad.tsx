import React from "react";
import {
  KEY_0,
  KEY_1,
  KEY_2,
  KEY_3,
  KEY_4,
  KEY_5,
  KEY_6,
  KEY_7,
  KEY_8,
  KEY_9,
  KEY_SEPARATOR,
  KEY_OP_ADD,
  KEY_OP_SUBTRACT,
  KEY_OP_MULTIPLY,
  KEY_OP_DIVIDE,
  KEY_OP_PERCENT,
  KEY_OP_EQUALS,
  KEY_CTRL_AC,
  KEY_CTRL_C,
} from "../../helpers/keys";

import "./Keypad.css";

/**
 * Keypad is collection of keys (buttons) organized in grid. Click on each button
 * calls onKeyPunch callback method with related key as parameter.
 *
 *    +------------------------------+
 *    |                              |
 *    |  ( AC)  ( C )  ( % )  ( / )  |
 *    |                              |
 *    |  ( 7 )  ( 8 )  ( 9 )  ( * )  |
 *    |                              |
 *    |  ( 4 )  ( 5 )  ( 6 )  ( - )  |
 *    |                              |
 *    |  ( 1 )  ( 2 )  ( 3 )  ( + )  |
 *    |                              |
 *    |  (     0    )  ( . )  ( = )  |
 *    |                              |
 *    +------------------------------+
 */

type KeypadProps = {
  onKeyPunch: (key: string) => void;
};

function Keypad({ onKeyPunch }: KeypadProps) {
  return (
    <section className="keypad" aria-label="Keypad">
      {/* First row - [AC] [C] [%] [/] */}

      <button
        type="button"
        className="keypad-btn dark"
        onClick={() => onKeyPunch(KEY_CTRL_AC)}
      >
        AC
      </button>

      <button
        type="button"
        className="keypad-btn dark"
        onClick={() => onKeyPunch(KEY_CTRL_C)}
      >
        C
      </button>

      <button
        type="button"
        className="keypad-btn dark"
        onClick={() => onKeyPunch(KEY_OP_PERCENT)}
      >
        %
      </button>

      <button
        type="button"
        className="keypad-btn dark"
        onClick={() => onKeyPunch(KEY_OP_DIVIDE)}
      >
        รท
      </button>

      {/* Second row - [7] [8] [9] [*] */}

      <button
        type="button"
        className="keypad-btn"
        onClick={() => onKeyPunch(KEY_7)}
      >
        7
      </button>

      <button
        type="button"
        className="keypad-btn"
        onClick={() => onKeyPunch(KEY_8)}
      >
        8
      </button>

      <button
        type="button"
        className="keypad-btn"
        onClick={() => onKeyPunch(KEY_9)}
      >
        9
      </button>

      <button
        type="button"
        className="keypad-btn dark"
        onClick={() => onKeyPunch(KEY_OP_MULTIPLY)}
      >
        ร
      </button>

      {/* Third row - [4] [5] [6] [-] */}

      <button
        type="button"
        className="keypad-btn"
        onClick={() => onKeyPunch(KEY_4)}
      >
        4
      </button>

      <button
        type="button"
        className="keypad-btn"
        onClick={() => onKeyPunch(KEY_5)}
      >
        5
      </button>

      <button
        type="button"
        className="keypad-btn"
        onClick={() => onKeyPunch(KEY_6)}
      >
        6
      </button>

      <button
        type="button"
        className="keypad-btn dark"
        onClick={() => onKeyPunch(KEY_OP_SUBTRACT)}
      >
        -
      </button>

      {/* Fourth row - [1] [2] [3] [+] */}

      <button
        type="button"
        className="keypad-btn"
        onClick={() => onKeyPunch(KEY_1)}
      >
        1
      </button>

      <button
        type="button"
        className="keypad-btn"
        onClick={() => onKeyPunch(KEY_2)}
      >
        2
      </button>

      <button
        type="button"
        className="keypad-btn"
        onClick={() => onKeyPunch(KEY_3)}
      >
        3
      </button>

      <button
        type="button"
        className="keypad-btn dark"
        onClick={() => onKeyPunch(KEY_OP_ADD)}
      >
        +
      </button>

      {/* Fifth row - [ 0 ] [.] [=] */}

      <button
        type="button"
        className="keypad-btn wide"
        onClick={() => onKeyPunch(KEY_0)}
      >
        0
      </button>

      <button
        type="button"
        className="keypad-btn"
        onClick={() => onKeyPunch(KEY_SEPARATOR)}
      >
        .
      </button>

      <button
        type="button"
        className="keypad-btn primary"
        autoFocus
        onClick={() => onKeyPunch(KEY_OP_EQUALS)}
      >
        =
      </button>
    </section>
  );
}

export default React.memo(Keypad);
