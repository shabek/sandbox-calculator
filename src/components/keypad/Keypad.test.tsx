import { vi } from "vitest";
import { getByRole, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import {
  KEY_0,
  KEY_1,
  KEY_2,
  KEY_3,
  KEY_4,
  KEY_5,
  KEY_6,
  KEY_7,
  KEY_8,
  KEY_9,
  KEY_SEPARATOR,
  KEY_OP_ADD,
  KEY_OP_SUBTRACT,
  KEY_OP_MULTIPLY,
  KEY_OP_DIVIDE,
  KEY_OP_PERCENT,
  KEY_OP_EQUALS,
  KEY_CTRL_AC,
  KEY_CTRL_C,
} from "../../helpers/keys";

import Keypad from "./Keypad";

describe("<Keypad />", () => {
  test("renders", () => {
    const { asFragment } = render(<Keypad onKeyPunch={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("click on button calls onKeyPunch callback with correct key", async () => {
    const user = userEvent.setup();

    const onPunchMock = vi.fn(() => {});
    render(<Keypad onKeyPunch={onPunchMock} />);

    const keypad = screen.getByLabelText("Keypad");
    expect(keypad).toBeInTheDocument();

    await user.click(getByRole(keypad, "button", { name: "AC" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_CTRL_AC);

    await user.click(getByRole(keypad, "button", { name: "C" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_CTRL_C);

    await user.click(getByRole(keypad, "button", { name: "%" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_OP_PERCENT);

    await user.click(getByRole(keypad, "button", { name: "รท" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_OP_DIVIDE);

    await user.click(getByRole(keypad, "button", { name: "7" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_7);

    await user.click(getByRole(keypad, "button", { name: "8" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_8);

    await user.click(getByRole(keypad, "button", { name: "9" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_9);

    await user.click(getByRole(keypad, "button", { name: "ร" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_OP_MULTIPLY);

    await user.click(getByRole(keypad, "button", { name: "4" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_4);

    await user.click(getByRole(keypad, "button", { name: "5" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_5);

    await user.click(getByRole(keypad, "button", { name: "6" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_6);

    await user.click(getByRole(keypad, "button", { name: "-" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_OP_SUBTRACT);

    await user.click(getByRole(keypad, "button", { name: "1" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_1);

    await user.click(getByRole(keypad, "button", { name: "2" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_2);

    await user.click(getByRole(keypad, "button", { name: "3" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_3);

    await user.click(getByRole(keypad, "button", { name: "+" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_OP_ADD);

    await user.click(getByRole(keypad, "button", { name: "0" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_0);

    await user.click(getByRole(keypad, "button", { name: "." }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_SEPARATOR);

    await user.click(getByRole(keypad, "button", { name: "=" }));
    expect(onPunchMock).toHaveBeenCalledWith(KEY_OP_EQUALS);
  });
});
