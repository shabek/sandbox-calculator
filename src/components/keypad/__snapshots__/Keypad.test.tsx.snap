// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Keypad /> > renders 1`] = `
<DocumentFragment>
  <section
    aria-label="Keypad"
    class="keypad"
  >
    <button
      class="keypad-btn dark"
      type="button"
    >
      AC
    </button>
    <button
      class="keypad-btn dark"
      type="button"
    >
      C
    </button>
    <button
      class="keypad-btn dark"
      type="button"
    >
      %
    </button>
    <button
      class="keypad-btn dark"
      type="button"
    >
      รท
    </button>
    <button
      class="keypad-btn"
      type="button"
    >
      7
    </button>
    <button
      class="keypad-btn"
      type="button"
    >
      8
    </button>
    <button
      class="keypad-btn"
      type="button"
    >
      9
    </button>
    <button
      class="keypad-btn dark"
      type="button"
    >
      ร
    </button>
    <button
      class="keypad-btn"
      type="button"
    >
      4
    </button>
    <button
      class="keypad-btn"
      type="button"
    >
      5
    </button>
    <button
      class="keypad-btn"
      type="button"
    >
      6
    </button>
    <button
      class="keypad-btn dark"
      type="button"
    >
      -
    </button>
    <button
      class="keypad-btn"
      type="button"
    >
      1
    </button>
    <button
      class="keypad-btn"
      type="button"
    >
      2
    </button>
    <button
      class="keypad-btn"
      type="button"
    >
      3
    </button>
    <button
      class="keypad-btn dark"
      type="button"
    >
      +
    </button>
    <button
      class="keypad-btn wide"
      type="button"
    >
      0
    </button>
    <button
      class="keypad-btn"
      type="button"
    >
      .
    </button>
    <button
      class="keypad-btn primary"
      type="button"
    >
      =
    </button>
  </section>
</DocumentFragment>
`;
